docs/gallery/tutorial/autogen/materials_science_ase.py

```diff
@@ -0,0 +1,316 @@
+# The process involves several steps:
+#
+# 1.  **Relax** the initial atomic structure to its lowest-energy state.
+# 2.  **Strain** the relaxed structure by applying a series of scaling factors.
+# 3.  **Calculate** the total energy and volume for each strained structure.
+# 4.  **Fit** the resulting energy-volume data to an EOS model to find properties like the equilibrium volume and bulk modulus.
+#
+# This workflow perfectly demonstrates how to handle loops and a dynamic number of calculations.
+
+from ase.calculators.emt import EMT
+from ase.optimize import BFGS
+
+
+@task()
+def relax_structure(atoms: Atoms) -> Atoms:
+    """Relax the atomic structure to its minimum energy configuration using ASE."""
+    atoms.calc = EMT()
+    optimizer = BFGS(atoms)
+    optimizer.run(fmax=0.01)
+    return atoms
+
+
+@task(
+    outputs={
+        "scaled_structures": {
+            "identifier": "workgraph.namespace",
+            "metadata": {"dynamic": True},
+        }
+    }
+)
+def create_strained_structures(atoms: Atoms, scales: list) -> dict:
+    """Generate a series of strained structures from a list of scaling factors."""
+    scaled_structures = {}
+    for i, scale in enumerate(scales):
+        strained_atoms = atoms.copy()
+        strained_atoms.set_cell(atoms.get_cell() * scale, scale_atoms=True)
+        # Each structure gets a unique key, like "strain_0", "strain_1", etc.
+        scaled_structures[f"strain_{i}"] = strained_atoms
+    return {"scaled_structures": scaled_structures}
+
+
+# %%
+# .. important::
+#    **Handling dynamic outputs with namespaces** ✨
+#
+#    In our EOS workflow, the ``create_strained_structures`` task generates a *variable* number of outputs. How do we track each one?
+#
+#    The solution is a **dynamic output namespace**. By adding the ``outputs={...}`` argument to the ``@task`` decorator, we tell AiiDA-WorkGraph:
+#    "The ``scaled_structures`` output is a dictionary, but I want you to treat *each key-value pair* as a separate, named output port (e.g., ``scaled_structures.strain_0``, ``scaled_structures.strain_1``)."
+#
+#    This is crucial for two reasons:
+#
+#    1.  **Full Provenance**: Every strained structure becomes an individual, tracked AiiDA node.
+#        This gives you a crystal-clear history of your entire calculation, with no "black boxes".
+#    2.  **Database Integrity**: It avoids saving a list of ``Atoms`` objects as a single, opaque file (e.g., using pickle, disabled by default).
+#        Instead, each structure is a queryable, first-class citizen in the AiiDA database.
+
+
+@task()
+def calculate_energy_and_volume(atoms: Atoms) -> dict:
+    """Calculate the energy and volume for a single atomic structure."""
+    atoms.calc = EMT()
+    atoms.get_potential_energy()
+    results = {
+        "energy": atoms.calc.results["energy"],
+        "volume": atoms.get_volume(),
+    }
+    return results
```

Comment:
```suggestion
    return {
        "energy": atoms.calc.results["energy"],
        "volume": atoms.get_volume(),
    }
```

I guess since the task does not explicitly declares its outputs, this is treated as a dictionary under the default `result` socket?

Reply:
Yes.